contracts/Items/DEFYForge.sol

```diff
@@ -0,0 +1,405 @@
+
+    // Event indexed by operative address and forgeJobId, includes burned tokens.
+    event CreateForgeJob(
+        address indexed operativeAddress,
+        uint256 indexed forgeJobId,
+        uint256 indexed blueprintId,
+        uint256[] inputLootIds,
+        uint256[] inputLootAmounts
+    );
+
+    // Event indexed by operative address and forgeJobId, includes minted tokens.
+    event CompleteForgeJob(
+        address indexed operativeAddress,
+        uint256 indexed forgeJobId,
+        uint256 indexed blueprintId,
+        uint256 outputLootId
+    );
+
+    // Event indexed by operative address and forgeJobId.
+    event CancelForgeJob(
+        address indexed operativeAddress,
+        uint256 indexed forgeJobId,
+        uint256[] remintedLootIds,
+        uint256[] remintedLootAmounts
+    );
+
+    // Event indexed by operative address and forgeJobId.
+    event FailForgeJob(
+        address indexed operativeAddress,
+        uint256 indexed forgeJobId
+    );
+
+    constructor() {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    /**
+     * @dev Create a forge job for an operative.
+     *      Burns the input loots.
+     * @return the forge job id.
+     */
+    function createForgeJob(
+        IDEFYLoot lootContract,
+        address operativeAddress,
+        uint256[] calldata inputLootIds,
+        uint256[] calldata inputLootAmounts,
+        uint256 duration,
+        uint256 blueprintId,
+        uint256 outputLootId
+    ) public onlyRole(FORGER_ROLE) returns (uint256) {
+        // require call is made to a valid loot contract
+        require(
+            validLootContracts[lootContract] == true,
+            "DEFYForge: Loot contract not valid"
+        );
+
+        // require input arrays are not null
+        require(
+            inputLootAmounts.length != 0 && inputLootIds.length != 0,
+            "DEFYForge: Invalid input loots"
+        );
+
+        // require input loots ids and amount are the same
+        require(
+            inputLootAmounts.length == inputLootIds.length,
+            "DEFYForge: All arrays must be the same length"
+        );
+
+        // require operative has tokens available
+        for (uint256 i = 0; i < inputLootIds.length; i++) {
+            require(
+                lootContract.balanceOf(operativeAddress, inputLootIds[i]) >=
+                    inputLootAmounts[i],
+                "DEFYForge: Operative does not have suffient loots"
+            );
+        }
+
+        // create print job record
+        uint256 forgeJobId = _forgeJobIds.current();
+
+        // increment _forgeJobId counter
+        _forgeJobIds.increment();
+
+        // Add new forge job to storage mapping
+        forgeJobs[forgeJobId] = ForgeJob(
+            lootContract,
+            operativeAddress,
+            inputLootIds,
+            inputLootAmounts,
+            timeNow(),
+            timeNow() + duration,
+            blueprintId,
```

Comment:
Hey, could you please explain why validation of the inputs is not needed? I mean the `inputLootIds` and `inputLootAmounts` are just taken from the arguments. Do we need to validate that for specific  `blueprintId` `inputLootIds` and `inputLootAmounts` are correct? Or this is not relevant because only authorised address can execute this smart contract, so paremeters are controlled by our code on the BE?
And does `blueprintId` correspond to the DefinitionId in the BE domain?

Reply:
The `inputLootIds` and `inputLootAmounts` are to be checked on our backend, as I think that not all the blueprints are going to be on-chain.
The backend logic should be checking that the inputs are correct for the blueprint, and that the operative has the items. 
Yes a `bluprintId` corresponds to a DefinitionId and BE domain entity.

Reply:
Thanks, it makes sense.
I have one concern about the data types for the Item/Blueprint ids. It's `uint256` now. In the BE code we have them as `Guid` that is an 128 bit integer basically but I'm not sure if it can be mapped to `uint256`. Furthermore it's quite possible that the id of an ItemDefinition is about to be `string` in case if we'll make Contentful the source of truth for the item definitions. Is it possible to have itemIds (ones that are mapped to the Definitions on the BE) to have string Ids?